File: nodejs/test/e2e/tool_results.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { describe, expect, it } from "vitest";
import { z } from "zod";
import type { SessionEvent, ToolResultObject } from "../../src/index.js";
import { approveAll, defineTool } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext";

describe("Tool Results", async () => {
    const { copilotClient: client, openAiEndpoint } = await createSdkTestContext();

    it("should handle structured ToolResultObject from custom tool", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            tools: [
                defineTool("get_weather", {
                    description: "Gets weather for a city",
                    parameters: z.object({
                        city: z.string(),
                    }),
                    handler: ({ city }): ToolResultObject => ({
                        textResultForLlm: `The weather in ${city} is sunny and 72°F`,
                        resultType: "success",
                    }),
                }),
            ],
        });

        const assistantMessage = await session.sendAndWait({
            prompt: "What's the weather in Paris?",
        });

        const content = assistantMessage?.data.content ?? "";
        expect(content).toMatch(/sunny|72/i);

        await session.disconnect();
    });

    it("should handle tool result with failure resultType", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            tools: [
                defineTool("check_status", {
                    description: "Checks the status of a service",
                    handler: (): ToolResultObject => ({
                        textResultForLlm: "Service unavailable",
                        resultType: "failure",
                        error: "API timeout",
                    }),
                }),
            ],
        });

        const assistantMessage = await session.sendAndWait({
            prompt: "Check the status of the service using check_status. If it fails, say 'service is down'.",
        });

        const failureContent = assistantMessage?.data.content ?? "";
        expect(failureContent).toMatch(/service is down/i);

        await session.disconnect();
    });

    it("should pass validated Zod parameters to tool handler", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            tools: [
                defineTool("calculate", {
                    description: "Calculates a math expression",
                    parameters: z.object({
                        operation: z.enum(["add", "subtract", "multiply"]),
                        a: z.number(),
                        b: z.number(),
                    }),
                    handler: ({ operation, a, b }) => {
                        expect(typeof a).toBe("number");
                        expect(typeof b).toBe("number");
                        switch (operation) {
                            case "add":
                                return String(a + b);
                            case "subtract":
                                return String(a - b);
                            case "multiply":
                                return String(a * b);
                        }
                    },
                }),
            ],
        });

        const assistantMessage = await session.sendAndWait({
            prompt: "Use calculate to add 17 and 25",
        });

        expect(assistantMessage?.data.content).toContain("42");

        await session.disconnect();
    });

    it("should preserve toolTelemetry and not stringify structured results for LLM", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            tools: [
                defineTool("analyze_code", {
                    description: "Analyzes code for issues",
                    parameters: z.object({
                        file: z.string(),
                    }),
                    handler: ({ file }): ToolResultObject => ({
                        textResultForLlm: `Analysis of ${file}: no issues found`,
                        resultType: "success",
                        toolTelemetry: {
                            metrics: { analysisTimeMs: 150 },
                            properties: { analyzer: "eslint" },
                        },
                    }),
                }),
            ],
        });

        const events: SessionEvent[] = [];
        session.on((event) => events.push(event));

        const assistantMessage = await session.sendAndWait({
            prompt: "Analyze the file main.ts for issues.",
        });

        expect(assistantMessage?.data.content).toMatch(/no issues/i);

        // Verify the LLM received just textResultForLlm, not stringified JSON
        const traffic = await openAiEndpoint.getExchanges();
        const lastConversation = traffic[traffic.length - 1]!;
        const toolResults = lastConversation.request.messages.filter(
            (m: { role: string }) => m.role === "tool"
        );
        expect(toolResults.length).toBe(1);
        expect(toolResults[0]!.content).not.toContain("toolTelemetry");
        expect(toolResults[0]!.content).not.toContain("resultType");

        // Verify tool.execution_complete event fires for this tool call
        const toolCompletes = events.filter((e) => e.type === "tool.execution_complete");
        expect(toolCompletes.length).toBeGreaterThanOrEqual(1);
        const completeEvent = toolCompletes[0]!;
        expect(completeEvent.data.success).toBe(true);
        // When the server preserves the structured result, toolTelemetry should
        // be present and non-empty (not the {} that results from stringification).
        if (completeEvent.data.toolTelemetry) {
            expect(Object.keys(completeEvent.data.toolTelemetry).length).toBeGreaterThan(0);
        }

        await session.disconnect();
    });
});

File: go/internal/e2e/tool_results_test.go
package e2e

import (
	"strings"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

func TestToolResults(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("should handle structured toolresultobject from custom tool", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		type WeatherParams struct {
			City string `json:"city" jsonschema:"City name"`
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Tools: []copilot.Tool{
				copilot.DefineTool("get_weather", "Gets weather for a city",
					func(params WeatherParams, inv copilot.ToolInvocation) (copilot.ToolResult, error) {
						return copilot.ToolResult{
							TextResultForLLM: "The weather in " + params.City + " is sunny and 72°F",
							ResultType:       "success",
						}, nil
					}),
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		_, err = session.Send(t.Context(), copilot.MessageOptions{Prompt: "What's the weather in Paris?"})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		answer, err := testharness.GetFinalAssistantMessage(t.Context(), session)
		if err != nil {
			t.Fatalf("Failed to get assistant message: %v", err)
		}

		content := ""
		if answer.Data.Content != nil {
			content = *answer.Data.Content
		}
		if !strings.Contains(strings.ToLower(content), "sunny") && !strings.Contains(content, "72") {
			t.Errorf("Expected answer to mention sunny or 72, got %q", content)
		}

		if err := session.Disconnect(); err != nil {
			t.Errorf("Failed to disconnect session: %v", err)
		}
	})

	t.Run("should handle tool result with failure resulttype", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Tools: []copilot.Tool{
				{
					Name:        "check_status",
					Description: "Checks the status of a service",
					Handler: func(inv copilot.ToolInvocation) (copilot.ToolResult, error) {
						return copilot.ToolResult{
							TextResultForLLM: "Service unavailable",
							ResultType:       "failure",
							Error:            "API timeout",
						}, nil
					},
				},
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		_, err = session.Send(t.Context(), copilot.MessageOptions{
			Prompt: "Check the status of the service using check_status. If it fails, say 'service is down'.",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		answer, err := testharness.GetFinalAssistantMessage(t.Context(), session)
		if err != nil {
			t.Fatalf("Failed to get assistant message: %v", err)
		}

		content := ""
		if answer.Data.Content != nil {
			content = *answer.Data.Content
		}
		if !strings.Contains(strings.ToLower(content), "service is down") {
			t.Errorf("Expected 'service is down', got %q", content)
		}

		if err := session.Disconnect(); err != nil {
			t.Errorf("Failed to disconnect session: %v", err)
		}
	})

	t.Run("should preserve tooltelemetry and not stringify structured results for llm", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		type AnalyzeParams struct {
			File string `json:"file" jsonschema:"File to analyze"`
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Tools: []copilot.Tool{
				copilot.DefineTool("analyze_code", "Analyzes code for issues",
					func(params AnalyzeParams, inv copilot.ToolInvocation) (copilot.ToolResult, error) {
						return copilot.ToolResult{
							TextResultForLLM: "Analysis of " + params.File + ": no issues found",
							ResultType:       "success",
							ToolTelemetry: map[string]any{
								"metrics":    map[string]any{"analysisTimeMs": 150},
								"properties": map[string]any{"analyzer": "eslint"},
							},
						}, nil
					}),
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		_, err = session.Send(t.Context(), copilot.MessageOptions{Prompt: "Analyze the file main.ts for issues."})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		answer, err := testharness.GetFinalAssistantMessage(t.Context(), session)
		if err != nil {
			t.Fatalf("Failed to get assistant message: %v", err)
		}

		content := ""
		if answer.Data.Content != nil {
			content = *answer.Data.Content
		}
		if !strings.Contains(strings.ToLower(content), "no issues") {
			t.Errorf("Expected 'no issues', got %q", content)
		}

		// Verify the LLM received just textResultForLlm, not stringified JSON
		traffic, err := ctx.GetExchanges()
		if err != nil {
			t.Fatalf("Failed to get exchanges: %v", err)
		}

		lastConversation := traffic[len(traffic)-1]
		var toolResults []testharness.ChatCompletionMessage
		for _, msg := range lastConversation.Request.Messages {
			if msg.Role == "tool" {
				toolResults = append(toolResults, msg)
			}
		}

		if len(toolResults) != 1 {
			t.Fatalf("Expected 1 tool result, got %d", len(toolResults))
		}
		if strings.Contains(toolResults[0].Content, "toolTelemetry") {
			t.Error("Tool result content should not contain 'toolTelemetry'")
		}
		if strings.Contains(toolResults[0].Content, "resultType") {
			t.Error("Tool result content should not contain 'resultType'")
		}

		if err := session.Disconnect(); err != nil {
			t.Errorf("Failed to disconnect session: %v", err)
		}
	})
}
